### PR TITLE
Adding quote_style option. Fixes #293.

### LINKED ISSUE #293: Add option to preserve quote style
I do a bunch of weird stuff with JSON in one of my projects, with lots of weird escaping/unescaping. I've had to disable `grunt-contrib-uglify` because it changes all of my quotes around and breaks everything.

HOWEVER, [this recent change to UglifyJS](https://github.com/mishoo/UglifyJS2/commit/fbbaa42ee55a7f753f7cab9b1a905ccf73cf26d5) grants the ability to preserve quotes. It would be *phenomenal* to have it as an option in `grunt-contrib-uglify`.

### PR DESCRIPTION
**Pull request by aendrew**
See #293. Also includes new test fixture. Two notes:

1. I didn't write a test for the default case (`-q 0`). Please let me know if I should do so.
1. I used the same semi-descriptive flags that `uglifyjs` does — i.e., `0`, `1`, `2`, `3`. Please let me know if I should change it to something more intuitive like `single`, `double`, `original`, `default`.

### CODE REVIEW

**Comment by aendrew** [comment]
Not sure why Travis is failing, I haven't touched the failing tests. Note that the same tests failed upon cloning to local.

**Comment by XhmikosR** [comment]
Travis failures are unrelated, I'm on it.

File: Gruntfile.js [hunk 1]
```diff
@@ -338,6 +362,9 @@ module.exports = function(grunt) {
     'uglify:sourcemapin_sources',
     'uglify:expression_json',
     'uglify:expression_js',
+    'uglify:quotes_single',
+    'uglify:quotes_double',
+    'uglify:quotes_original',
```
Comment (XhmikosR) [inline Gruntfile.js]: @vladikoff: how about we make this just `uglify` instead of specifying each subtarget?

File: docs/uglify-options.md [hunk 2]
```diff
@@ -139,3 +139,13 @@ Type: `Boolean`
 Default: false
 
 Pass this flag if you don't care about full compliance with Internet Explorer 6-8 quirks.
+
+## quote_style
+Type: `Integer`
```
Comment (XhmikosR) [inline docs/uglify-options.md]: Needs two trailing spaces for a line break.

File: tasks/lib/uglify.js [hunk 3]
```diff
@@ -203,6 +203,10 @@ exports.init = function(grunt) {
       outputOptions.ascii_only = options.ASCIIOnly;
     }
 
+    if (options.quote_style !== undefined) {
```
Comment (XhmikosR) [inline tasks/lib/uglify.js]: Maybe `_.isUndefined(options.quote_style)`?

**Comment by aendrew** [comment]
Apols for the delay, will update the commit first thing tomorrow!

**Comment by aendrew** [comment]
@XhmikosR PR updated. Also changed the "typeof varName !== 'undefined" stanzas starting at `tasks/lib/uglify.js` to use `_.isUndefined()` for consistency, hope that's okay.

**Comment by XhmikosR** [comment]
@aendrew: please rebase and sort https://github.com/gruntjs/grunt-contrib-uglify/pull/306#discussion_r26437557 too.

**Comment by aendrew** [comment]
Hey @XhmikosR,
I fixed the merge conflicts (and added two spaces) by merging master into my branch and then using git mergetool; though I'm totally lost as to how I should rebase now that there have been commits between when my initial and updated commits. Normally I just do `git rebase -i HEAD~2`, but that'll rewrite the 8 or 9 commits since then. Apologies for the utterly newbie question, but how do I rebase such that my commits can be squashed and the other history is preserved? Thanks!

**Comment by XhmikosR** [comment]
Plain git rebase -i  should do the job. Otherwise 0ing me and I'll do it
myself.

**Comment by aendrew** [comment]
I think I've got it now — I merged upstream/master into local master, renamed my local working branch to "temp", checked out a new `293_adding_quote_style_option` branch from `master`, did a `git merge --squash temp` on `293...` and then committed afresh, force-pushing to my branch. 

Did I totally overthink this? Should I have just, in the beginning, pulled upstream master into my current branch, merged, resolved merge conflict, committed, and then squashed my merge commit into the one before it?

Thanks for taking the time to explain this to me — I should really know this, given how many open source projects I contribute to... :sweat_smile: Thanks!

File: docs/uglify-options.md [hunk 4]
```diff
@@ -167,3 +167,14 @@ A string that is a path to a JSON cache file that uglify will create and use to
 multiple runs of uglify. Note: this generated file uses the same JSON format as the `exceptionsFiles` files.
 
 
+
+## quote_style
+Type: `Integer`
```
Comment (XhmikosR) [inline docs/uglify-options.md]: Needs 2 trailing spaces for a line break.

**Comment by XhmikosR** [comment]
It looks fine now. The only downside the way you did it is that the commit date changed, but it's not important.

**Comment by aendrew** [comment]
@XhmikosR I think I've finally fixed the whitespace issue, now, finally. :sweat_smile:

**Comment by XhmikosR** [comment]
Nope :tongue:

File: Gruntfile.js [hunk 5]
```diff
@@ -326,6 +326,30 @@ module.exports = function(grunt) {
           mangleProperties: true,
           nameCache: 'tmp/uglify_name_cache.json'
         }
+      },
+      quotes_single: {
+          files: {
```
Comment (XhmikosR) [inline Gruntfile.js]: Oh, also, please use 2 spaces for indentation consistently.

File: README.md [hunk 6]
```diff
@@ -188,7 +188,18 @@ Type: `Boolean`
 Default: false
 
 Pass this flag if you don't care about full compliance with Internet Explorer 6-8 quirks.
-
+  
+#### quote_style
```
Comment (XhmikosR) [inline README.md]: Finally, `quoteStyle` seems consistent with the scheme that's already being used, i.e. camelcase.

**Comment by XhmikosR** [comment]
@aendrew: those last comments should be it. Thanks for the PR :+1:

**Comment by aendrew** [comment]
@XhmikosR Is the whitespace in the rest of README.md fine? Because I have invisibles on and the only lines with two spaces and a newline are my lines.

To confirm I'm understanding correctly — on lines that are only newline, you want `\s\s\n`, not simply `\n`?

Updating rest of PR now.

**Comment by XhmikosR** [comment]
In markdown two trailing spaces means a line break, so yeah.

**Comment by aendrew** [comment]
Yay! Thanks! :+1: :shipit:

**Comment by XhmikosR** [comment]
I added the line breaks myself in https://github.com/gruntjs/grunt-contrib-uglify/commit/672d94aae0097b8569c243a5a5164bb8f01c4bc7

**Comment by aendrew** [comment]
@XhmikosR Thanks for doing that, not sure why I was having such issues — even after disabling the Atom Whitespace plugin.

**Comment by XhmikosR** [comment]
New stuff is nice and all, but I'm old school and I use [my Notepad2 fork](https://github.com/XhmikosR/notepad2-mod)  :)

**Pull request merged**